satpy/writers/cf_writer.py

```diff
@@ -251,32 +180,25 @@ def area2lonlat(dataarray):
                         name='latitude')
     dataarray['longitude'] = lons
     dataarray['latitude'] = lats
-    return [dataarray]
 
 
 def area2gridmapping(dataarray):
     """Convert an area to at CF grid mapping."""
     area = dataarray.attrs['area']
-    attrs = create_grid_mapping(area)
-    if attrs is not None and 'name' in attrs.keys() and attrs['name'] != "proj4":
-        dataarray.attrs['grid_mapping'] = attrs['name']
-        name = attrs['name']
-    else:
-        # Handle the case when the projection cannot be converted to a standard CF representation or this has not
-        # been implemented yet.
-        dataarray.attrs['grid_proj4'] = area.proj4_string
-        name = "proj4"
-    return [dataarray, xr.DataArray(0, attrs=attrs, name=name)]
+    gmapping_var_name, attrs = create_grid_mapping(area)
+    dataarray.attrs['grid_mapping'] = gmapping_var_name
+    return xr.DataArray(0, attrs=attrs, name=gmapping_var_name)
 
 
 def area2cf(dataarray, strict=False):
     """Convert an area to at CF grid mapping or lon and lats."""
     res = []
     dataarray = dataarray.copy(deep=True)
     if isinstance(dataarray.attrs['area'], SwathDefinition) or strict:
-        res = area2lonlat(dataarray)
+        # modifies dataarray in-place
+        area2lonlat(dataarray)
     if isinstance(dataarray.attrs['area'], AreaDefinition):
-        res.extend(area2gridmapping(dataarray))
+        res.append(area2gridmapping(dataarray))
```

Comment:
Same here, can we have `area2gridmapping` also return a modified copy of the array ? In this case we probably just need a shallow copy as we just add an attribute.